File: src/jsonrpc/eth/getBalance/eth_getBalance.js
/**
 * @fileoverview eth_getBalance JSON-RPC method
 */

/**
 * @typedef {import('../../types/index.js').AddressType} Address
 * @typedef {import('../../types/index.js').Hash} Hash
 * @typedef {import('../../types/index.js').Quantity} Quantity
 * @typedef {import('../../types/index.js').BlockTag} BlockTag
 * @typedef {import('../../types/index.js').BlockSpec} BlockSpec
 * @typedef {import('../../../provider/types.js').RequestArguments} RequestArguments
 */

/**
 * Returns the balance of the account of given address.
 *
 * @example
 * Address: "0xfe3b557e8fb62b89f4916b721be55ceb828dbd73"
 * Block: "latest"
 * Result: "0x1cfe56f3795885980000"
 *
 * Implements the `eth_getBalance` JSON-RPC method.
 */

/** The JSON-RPC method name */
export const method = "eth_getBalance";

/**
 * Result for `eth_getBalance`
 *
 * hex encoded unsigned integer
 *
 * @typedef {Quantity} Result
 */

/**
 * Creates an eth_getBalance JSON-RPC request
 *
 * @param {Address | `0x${string}`} address - Hex encoded address
 * @param {BlockSpec} [block='latest'] - Block number, tag, or block hash
 * @returns {RequestArguments}
 *
 * @example
 * ```js
 * const request = GetBalanceRequest('0x742d35Cc6634C0532925a3b844Bc9e7595f0bEb0', 'latest')
 * // { jsonrpc: '2.0', method: 'eth_getBalance', params: [...], id: null }
 * ```
 */
export function GetBalanceRequest(address, block = "latest") {
	return { method, params: [address, block] };
}

File: packages/voltaire-ts/src/jsonrpc/eth/getBalance/eth_getBalance.go
package eth

import (
	"encoding/json"
	"fmt"

	"github.com/ethereum/execution-apis/types"
)

// Returns the balance of the account of given address.
//
// Example:
// Address: "0xfe3b557e8fb62b89f4916b721be55ceb828dbd73"
// Block: "latest"
// Result: "0x1cfe56f3795885980000"
//
// Implements the eth_getBalance JSON-RPC method.

// Method is the JSON-RPC method name
const Method = "eth_getBalance"

// Params represents the parameters for eth_getBalance
type Params struct {
	// hex encoded address
	Address types.Address `json:"-"`
	// Block number, tag, or block hash
	Block types.BlockSpec `json:"-"`
}

// MarshalJSON implements json.Marshaler for Params.
// JSON-RPC 2.0 uses positional array parameters.
func (p Params) MarshalJSON() ([]byte, error) {
	return json.Marshal([]interface{}{
		p.Address,
		p.Block,
	})
}

// UnmarshalJSON implements json.Unmarshaler for Params.
func (p *Params) UnmarshalJSON(data []byte) error {
	var arr []json.RawMessage
	if err := json.Unmarshal(data, &arr); err != nil {
		return err
	}

	if len(arr) != 2 {
		return fmt.Errorf("expected 2 parameters, got %d", len(arr))
	}

	if err := json.Unmarshal(arr[0], &p.Address); err != nil {
		return fmt.Errorf("parameter 0 (Address): %w", err)
	}

	if err := json.Unmarshal(arr[1], &p.Block); err != nil {
		return fmt.Errorf("parameter 1 (Block): %w", err)
	}

	return nil
}

// Result represents the result for eth_getBalance
//
// hex encoded unsigned integer
type Result struct {
	Value types.Quantity `json:"-"`
}

// MarshalJSON implements json.Marshaler for Result.
func (r Result) MarshalJSON() ([]byte, error) {
	return json.Marshal(r.Value)
}

// UnmarshalJSON implements json.Unmarshaler for Result.
func (r *Result) UnmarshalJSON(data []byte) error {
	return json.Unmarshal(data, &r.Value)
}
